src/middlewared/middlewared/plugins/ldap.py

```diff
@@ -0,0 +1,641 @@
+import asyncio
+import enum
+import errno
+import grp
+import ldap
+import ldap.sasl
+import pwd
+import socket
+
+from ldap.controls import SimplePagedResultsControl
+from middlewared.schema import accepts, Bool, Dict, Int, List, Str
+from middlewared.service import job, private, ConfigService
+from middlewared.service_exception import CallError
+from middlewared.utils import run
+
+
+class DSStatus(enum.Enum):
+    """
+    Following items are used for cache entries indicating the status of the
+    Directory Service.
+    :FAULTED: Directory Service is enabled, but not HEALTHY.
+    :LEAVING: Directory Service is in process of stopping.
+    :JOINING: Directory Service is in process of starting.
+    :HEALTHY: Directory Service is enabled, and last status check has passed.
+    There is no "DISABLED" DSStatus because this is controlled by the "enable" checkbox.
+    This is a design decision to avoid conflict between the checkbox and the cache entry.
+    """
+    FAULTED = 1
+    LEAVING = 2
+    JOINING = 3
+    HEALTHY = 4
+
+
+class SSL(enum.Enum):
+    NOSSL = 'off'
+    USESSL = 'on'
+    USETLS = 'start_tls'
+
+
+class LDAPQuery(object):
+    def __init__(self, **kwargs):
+        super(LDAPQuery, self).__init__()
+        self.ldap = kwargs.get('conf')
+        self.logger = kwargs.get('logger')
+        self.hosts = kwargs.get('hosts')
+        self.pagesize = 1024
+        self._isopen = False
+        self._handle = None
+        self._rootDSE = None
+
+    def __enter__(self):
+        return self
+
+    def __exit__(self, typ, value, traceback):
+        if self._isopen:
+            try:
+                self._close()
+            except Exception as e:
+                raise CallError(f'Failed to close connection to LDAP server: {e}')
```

Comment:
You may have misunderstood me.
If an exception occurs within this context it will be silently skipped.
You need to re-raise it:

if typ is not None:
    raise